### PR TITLE
Gracefully handle invalid URL

### PR DESCRIPTION
**Pull request by maciejp-ro**
Some URLs that reach our cloudflare workers make Cloudflare's `new URL(event.request.url)` throw an exception. Specifically, broken percent-encoding (e.g. URL ending with a `%` character because it was truncated on its way) triggers this. There may be more cases when it happens.

Toucan instantiates `URL` during initialization, which throws an exception before exception-handling mechanism is even ready. This PR wraps `new URL` in a `try` clause and if it throws, naively splits by `?` character instead. This makes the initialization succeed and makes it possible to catch this exception when it happens in user code.

Currently we use internally one more patch on top of this, which adds a breadcrumb if parsing the URL failed. I'm not sure this should be forced on all users, but if you think it's a good idea, I can add it to this PR.

Unfortunately this is not reproducible in Node, but for internal tests of the complete worker, `new Request("garbage", …)` worked. Should I add a test case like this here?

### CODE REVIEW

**Review by robertcepa — COMMENTED** [review]
Test for `new Request("garbage", …)` would be great!

File: src/index.ts [hunk 1]
```diff
@@ -365,15 +365,28 @@ export default class Toucan {
...
+    try {
+      const url = new URL(request.url);
+      rv.url = `${url.protocol}//${url.hostname}${url.pathname}`;
+      rv.query_string = url.search;
+    } catch (e) {
+      // `new URL` failed, let's try to split URL the primitive way
+      const qi = request.url.indexOf('?');
+      if (qi < 0) {
+        // no query string
+        rv.url = request.url;
```
[the remark this answers is not in the document]
Comment (maciejp-ro) [reply src/index.ts:383]: It's not necessarily invalid. AFAIR, HTTP doesn't specify semantics of the URI (especially in the query string). `http://example.com/?something%` is not valid query string, not valid URL-encoding, but I believe it is still a valid URL if the server chooses to parse it.

Even if it's technically invalid, I believe it's useful to keep it in metadata – if the request throws an exception in worker code, the URL that triggered it is important information for debugging. Especially if the URL is suspicious to begin with.

**Comment by maciejp-ro** [comment]
Returning to this PR: added a test case, is there anything more required to merge it?

**Comment by maciejp-ro** [comment]
@robertcepa ping?

**Comment by robertcepa** [comment]
@maciejp-ro Sorry for the delay. This looks great. Thanks for contributing!

**Review by robertcepa — APPROVED** [review]

**Pull request merged**